pkg/networkservice/common/connect/server.go

```diff
@@ -21,247 +19,112 @@ package connect
 
 import (
 	"context"
-	"net/url"
-
-	"github.com/networkservicemesh/sdk/pkg/networkservice/core/trace"
-	"github.com/networkservicemesh/sdk/pkg/tools/grpcutils"
-	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
 
 	"github.com/golang/protobuf/ptypes/empty"
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/sdk/pkg/networkservice/common/clienturl"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/utils/inject/injecterror"
+	"github.com/networkservicemesh/sdk/pkg/tools/clientmap"
 )
 
-// Contain grpc connection and client connection associated
-type clientEntry struct {
-	clientURI       *url.URL
-	client          networkservice.NetworkServiceClient
-	cc              grpc.ClientConnInterface
-	ready           chan struct{} // Flag for connection is establishing
-	connections     map[string]*networkservice.Connection
-	err             error
-	closeConnection func() error
-}
-
-func (ce *clientEntry) markAsReady() {
-	close(ce.ready)
-}
-
 type connectServer struct {
-	ctx           context.Context
-	dialOptions   []grpc.DialOption
-	clientFactory func(ctx context.Context, conn grpc.ClientConnInterface) networkservice.NetworkServiceClient
-	clients       map[string]*clientEntry // key == url as string
-	connections   map[string]*url.URL     // Connection map is required to close using connection id.
-	executor      serialize.Executor
+	ctx               context.Context
+	clientFactory     func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient
+	clientDialOptions []grpc.DialOption
+	clientsByURL      clientmap.RefcountMap // key == clientURL.String()
+	clientsByID       clientmap.Map         // key == client connection ID
```

Comment:
I think to prevent race conditions we need to use mutexes.
```suggestion
	clients            map[string]*clientEntry // key == url as string		clientsByURL      clientmap.RefcountMap // key == clientURL.String()
	connections        map[string]*url.URL  
	lockers            map[string]sync.Mutex
```

Reply:
We actually don't need global mutexes to avoid race conditions (see comments below).  I *think* we produce correct behavior, even in the case of nearly simultaneous initial Requests for the same connection.ID.

Reply:
Definitely you are right looks like global mutex can be more costly for those cases.